snapcraft/commands/upload.py

```diff
@@ -107,6 +151,54 @@ def run(self, parsed_args):
         emit.message(message)
 
 
+def _validate_components(
+    provided_components: list[ComponentOption], snap_yaml: dict[str, Any]
+) -> None:
+    """Validate component data.
+
+    Validates:
+     - The components provided by the user match the components listed in the snap's metadata.
+     - All component files exist.
+
+    :param provided_components: A list of ComponentOptions provided by the user.
+    :param snap_yaml: The snap's metadata.
+    """
+    # get the names of the components provided by the user
+    provided_names = {
+        component.name for component in provided_components if component.name
+    }
+    emit.debug(f"Components provided by the user: {provided_names}")
+
+    if "components" in snap_yaml:
```

Comment:
can you add a followup to use the new model and load that instead to process this? Instead of looking at snap_yaml, look at a Snap object.

Reply:
hmm, the model is updated in this PR, so perhaps now?

Reply:
Agree with Sergio, this can be done against validated structured data instead of raw content.

Reply:
Good idea, added.